### ISSUE TITLE
Hessian of pairwise potential outputs NaNs

### ISSUE THREAD

**Issue by rsantet** [google/jax#18562]
### Description

When defining a custom Lennard-Jones potential, computing the gradient outputs the expected result, but the Hessian only outputs NaNs.

```python
import ml_collections
import jax.numpy as jnp
import jax
import numpy as np

# To store potential parameters
config = ml_collections.ConfigDict()

config.varepsilon = 1.0
config.r_cut = 2.5
config.sigma = 1.0
config.box_length = 5.5

# Make the potential smoother (with onset and cutoff)
reduced_r_cut_6 = (config.sigma / config.r_cut) ** 6
tmp = 1 / 0.8
tmp_6 = tmp**6

config.d = 2 * config.varepsilon * reduced_r_cut_6 * (0.5 * reduced_r_cut_6 - 1)
config.c = (
    12 * config.varepsilon * reduced_r_cut_6 * (reduced_r_cut_6 - 1) / config.r_cut
)
config.b = (
    12 * config.varepsilon * tmp * tmp_6 * (tmp_6 - 1) / config.sigma - config.c
) / (1.6 * config.sigma)
config.a = (
    config.b * (0.8 * config.sigma) ** 2
    + 2 * config.varepsilon * tmp_6 * (0.5 * tmp_6 - 1)
    + config.c * (0.8 * config.sigma - config.r_cut)
    - config.d
)

def LJ(r):
    outer = jnp.where(
        r > 0.8 * config.sigma,
        2
        * config.varepsilon
        * (config.sigma / r) ** 6
        * (0.5 * (config.sigma / r) ** 6 - 1)
        + config.c * (r - config.r_cut)
        - config.d,
        config.a - config.b * r**2,
    )
    U = jnp.where(r > config.r_cut, 0.0, outer)
    return jnp.array(U)

# Distance up to periodic boundary conditions
def displacement(x1, x2):
    dx = x1 - x2
    return dx - config.box_length * jnp.round(dx / config.box_length)

def distance(dR):
    dr = jnp.sum(dR**2, axis=-1)
    masked = jnp.where(dr > 0, dr, 0)
    return jnp.where(dr > 0, jnp.sqrt(masked), 0)

def metric(displacement):
    return lambda x1, x2: distance(displacement(x1, x2))

LJ_metric = metric(displacement)

def V(x):
    d = jax.vmap(jax.vmap(LJ_metric, (0, None)), (None, 0))
    dr = d(x, x)
    LJ_dr = LJ(dr)
    mask = jnp.float32(1.0) - jnp.eye(LJ_dr.shape[0], dtype=LJ_dr.dtype)
    LJ_dr = mask * LJ_dr

    potential_energy = jnp.array(jnp.sum(LJ_dr), dtype=LJ_dr.dtype) * jnp.float32(0.5)
    return potential_energy

np.random.seed(1234)
x = jnp.asarray(np.random.rand(26, 2) * config.box_length) # System is composed of 26 particles in dimension 2
nabla_V = jax.grad(V)
nabla2_V = jax.hessian(V)
V(x) # Outputs float
nabla_V(x) # Outputs (26,2) matrix, no NaNs
nabla2_V(x) # Outputs a (26,2,26,2) tensor full of NaNs
```

In my code I use the force `-nabla_V(x)` where a neural network modifies `x` before and after this call. Therefore, I get NaNs in the backward pass. Note that I used two `jnp.where` calls in the `distance` function to ensure that the gradients were not NaNs as well, I thought it would be enough for the Hessian to work as well.

### What jax/jaxlib version are you using?

jax v0.4.19, jaxlib v0.4.19

### Which accelerator(s) are you using?

CPU

### Additional system info

Linux

**Comment by jakevdp** [comment]
Thanks for the question! I ran your code and I'm not seeing any NaN values in the output: I tried in both JAX v0.4.19 and v0.4.20. Can you double-check the example code and confirm which JAX version you're using?

**Comment by rsantet** [comment]
Hello Jake, thank your for your time.
It seems it is working fine... Maybe I got NaNs from another session and it propagated somewhere in the variables cache. Anyhow, it works perfectly, so I am closing this, sorry for that !
